spec.md

```diff
@@ -522,6 +535,165 @@ the module.  The `c`{.firrtl} sub-field contained in the `b`{.firrtl} sub-field
+
+Reference ports are not expected to be synthesizable or representable in the
+target language and are omitted in the compiled design; they only exist
+at the FIRRTL level.
+
+Reference-type ports are statically routed through the design using
+the `forward`{.firrtl} and `export`{.firrtl} statements.
+
+There are two reference types, `Probe`{.firrtl} and `RWProbe`{.firrtl},
+described below.  These are used for indirect access to probes of the data
+underlying circuit constructs they originate from, captured using
+`probe`{.firrtl} expressions (see [@sec:probes]).
+
+`Probe`{.firrtl} types are read-only, and `RWProbe`{.firrtl} may be used with
+`force`{.firrtl} and related statements.
+Prefer the former as much as possible, as read-only probes impose fewer
+limitations and are more amenable to optimization.
+
+References must always be able to be statically traced to their target, or to
+an external module's output reference.
+
+### Probe Types
+
+Probe types are reference types used to access circuit elements' data remotely.
+
+There are two probe types: `Probe`{.firrtl} and `RWProbe`{.firrtl}.
+`RWProbe`{.firrtl} is a `Probe`{.firrtl} type, but not the other way around.
+
+Probe types are parametric over the type of data that they refer to, which is
+always passive (as defined in [@sec:passive-types]) even when the probed target
+is not (see [@sec:probes-and-passive-types]).
+Probe types cannot contain probe types.
+
+Conceptually probe types are single-direction views of the probed data-flow point.
+
+Examples:
+
+```firrtl
+Probe<UInt> ; readable reference to unsigned integer with inferred width
+RWProbe<{x: {y: UInt}}> ; readable and forceable reference to bundle
+```
+
+For details of how to read and write through probe types, see
+[@sec:reading-probe-references;@sec:force-and-release].
+
+All ports of probe type must be initialized with exactly one statement:
+an originating `export`{.firrtl} statement ([@sec:export]) using a
+`probe`{.firrtl} expression, or by forwarding an existing probe reference
+([@sec:forward]).
+
+Probe types are only allowed as part of module ports and may not appear
+anywhere else.
+
+Probes within vectors must be accessed using statically known index values;
+more generally, all expressions indexing to or through a probe type must not be
+dependent on dynamic values.
+
+Probe types may be specified as part of an external module (see
+[@sec:externally-defined-modules]), with the resolved referent
+optionally specified using `ref`{.firrtl} statements.
+
+### Input References
+
+Probe references are generally forwarded up the design hierarchy, being used to
+reach down into design internals from a higher point.
+As a result probe-type references are most often output ports, but may also be
+used on input ports internally, as described in this section.
+
+Input probe references are are allowed on internal modules, but they should be
+used with care: when probe references are resolved they must target a single
```

Comment:
The colon denotes the second half/clause explains the first half in a way not captured by the semicolon.  I'll rewrite to avoid the punctuation and related grammar uncertainty.